File: typescript/src/extras/langchain.ts
// LangChain.js integration for Asqav. Install: npm install @langchain/core.
// Default-on: await enableLangchainGovernance({ agentName: "my-chain" }) then chain.invoke.
// Manual: const h = await AsqavCallbackHandler.create({...}); chain.invoke(i, {callbacks:[h]}).

import { AsqavAdapter, raiseMissingPeer, type AsqavAdapterOptions } from "./_base.js";

// Local LangChain.js types so this module compiles even when the peer
// package is not installed. The actual base class is loaded lazily.

interface LCSerialized {
  id?: string[];
  name?: string;
  [key: string]: unknown;
}

interface LCGeneration {
  text?: string;
  [key: string]: unknown;
}

interface LCLLMResult {
  generations: LCGeneration[][];
  llmOutput?: Record<string, unknown> | null;
}

interface BaseCallbackHandlerLike {
  // The LangChain.js base class declares many optional handlers. We only
  // need ours to type-check; consumers receive a real subclass instance.
  name?: string;
}

type BaseCallbackHandlerCtor = new (...args: unknown[]) => BaseCallbackHandlerLike;

const MAX_PREVIEW = 200;

function truncate(s: string, max = MAX_PREVIEW): string {
  return s.length > max ? s.slice(0, max) : s;
}

// Lazily import BaseCallbackHandler from @langchain/core.
// Re-throws a clear error when the optional peer is missing.
async function loadBase(): Promise<BaseCallbackHandlerCtor> {
  try {
    // Dynamic specifier: the peer is optional (package.json peerDependenciesMeta).
    // @ts-ignore optional peer dependency
    const mod = (await import("@langchain/core/callbacks/base")) as unknown;
    const ctor = (mod as { BaseCallbackHandler?: BaseCallbackHandlerCtor })
      .BaseCallbackHandler;
    if (!ctor) {
      throw new Error("BaseCallbackHandler not exported from @langchain/core/callbacks/base");
    }
    return ctor;
  } catch (err) {
    raiseMissingPeer(
      "LangChain.js",
      "@langchain/core",
      "npm install @langchain/core",
      err,
    );
  }
}

export interface AsqavCallbackHandlerOptions extends AsqavAdapterOptions {
  /** Optional callback handler name; defaults to "asqav-callback-handler". */
  name?: string;
}

// Signs chain, tool, and LLM events via the Asqav governance pipeline.
// Use AsqavCallbackHandler.create() since the peer is loaded asynchronously.
export class AsqavCallbackHandler extends AsqavAdapter {
  /** Friendly name surfaced to LangChain's tracing UI. */
  public readonly name: string = "asqav-callback-handler";

  // Async factory: loads BaseCallbackHandler from the optional peer at runtime.
  static async create(
    opts: AsqavCallbackHandlerOptions,
  ): Promise<AsqavCallbackHandler & BaseCallbackHandlerLike> {
    const Base = await loadBase();

    class _AsqavCallbackHandlerImpl extends (Base as unknown as new () => object) {
      public readonly name: string;
      private readonly _asqav: AsqavCallbackHandler;

      constructor(asqav: AsqavCallbackHandler, name: string) {
        super();
        this._asqav = asqav;
        this.name = name;
      }

      handleChainStart(
        serialized: LCSerialized | null | undefined,
        inputs: Record<string, unknown>,
      ): void {
        // Guard: LCEL/RunnableLambda passes non-object inputs (str/arr/num).
        // Object.keys on a string yields index keys ["0","1",...], so guard first.
        const input_keys =
          inputs !== null &&
          typeof inputs === "object" &&
          !Array.isArray(inputs)
            ? Object.keys(inputs)
            : [];
        // serialized may be null on the LCEL/RunnableLambda path; fall back
        // to a stable "chain" label so the receipt is always signed.
        if (!serialized) {
          this._asqav._emit("chain:start", {
            chain: "chain",
            input_keys,
          });
          return;
        }
        const fallbackName = (serialized.id ?? ["unknown"]).slice(-1)[0];
        this._asqav._emit("chain:start", {
          chain: String(serialized.name ?? fallbackName),
          input_keys,
        });
      }

      handleChainEnd(outputs: Record<string, unknown>): void {
        this._asqav._emit("chain:end", {
          output_keys: Object.keys(outputs ?? {}),
        });
      }

      handleChainError(err: Error): void {
        this._asqav._emit("chain:error", {
          error_type: err?.name ?? "Error",
          error: truncate(String(err?.message ?? err)),
        });
      }

      handleToolStart(serialized: LCSerialized, input: string): void {
        const fallbackName = (serialized.id ?? ["unknown"]).slice(-1)[0];
        this._asqav._emit("tool:start", {
          tool: String(serialized.name ?? fallbackName),
          input: truncate(String(input ?? "")),
        });
      }

      handleToolEnd(output: unknown): void {
        const outStr = typeof output === "string" ? output : JSON.stringify(output);
        this._asqav._emit("tool:end", {
          output_type: typeof output,
          output_length: outStr?.length ?? 0,
        });
      }

      handleToolError(err: Error): void {
        this._asqav._emit("tool:error", {
          error_type: err?.name ?? "Error",
          error: truncate(String(err?.message ?? err)),
        });
      }

      handleLLMStart(serialized: LCSerialized, prompts: string[]): void {
        const fallbackName = (serialized.id ?? ["unknown"]).slice(-1)[0];
        this._asqav._emit("llm:start", {
          model: String(serialized.name ?? fallbackName),
          prompt_count: Array.isArray(prompts) ? prompts.length : 0,
        });
      }

      handleLLMEnd(result: LCLLMResult): void {
        const generationCount = (result?.generations ?? []).reduce(
          (acc, gens) => acc + (gens?.length ?? 0),
          0,
        );
        const ctx: Record<string, unknown> = { generation_count: generationCount };
        const usage = result?.llmOutput?.tokenUsage ?? result?.llmOutput?.token_usage;
        if (usage && typeof usage === "object") {
          ctx.token_usage = Object.fromEntries(
            Object.entries(usage as Record<string, unknown>).filter(
              ([, v]) => typeof v === "number",
            ),
          );
        }
        this._asqav._emit("llm:end", ctx);
      }

      handleLLMError(err: Error): void {
        this._asqav._emit("llm:error", {
          error_type: err?.name ?? "Error",
          error: truncate(String(err?.message ?? err)),
        });
      }
    }

    const wrapper = new AsqavCallbackHandler(opts);
    const impl = new _AsqavCallbackHandlerImpl(
      wrapper,
      opts.name ?? "asqav-callback-handler",
    );
    // Make the underlying signAction reachable through the wrapper.
    return impl as unknown as AsqavCallbackHandler & BaseCallbackHandlerLike;
  }

  /** Internal hook used by the inner LangChain subclass. */
  public _emit(actionType: string, context: Record<string, unknown>): void {
    this.signAction({ actionType, context });
  }
}

// -- Default-on entrypoint ---------------------------------------------------

interface LangchainContextModule {
  registerConfigureHook: (config: { contextVar?: string; inheritable?: boolean }) => void;
  setContextVariable: (name: string, value: unknown) => void;
}

// Lazily import registerConfigureHook/setContextVariable from @langchain/core/context.
async function loadContext(): Promise<LangchainContextModule> {
  try {
    // @ts-ignore optional peer dependency
    const mod = (await import("@langchain/core/context")) as unknown;
    const m = mod as Partial<LangchainContextModule>;
    if (!m.registerConfigureHook || !m.setContextVariable) {
      throw new Error("registerConfigureHook/setContextVariable not exported from @langchain/core/context");
    }
    return m as LangchainContextModule;
  } catch (err) {
    raiseMissingPeer(
      "LangChain.js",
      "@langchain/core",
      "npm install @langchain/core",
      err,
    );
  }
}

const ASQAV_LC_CONTEXT_VAR = "asqav_langchain_handler";
let hookRegistered = false;

// Turn on default-on Asqav governance for LangChain.js. Registers a configure
// hook once and binds the handler to a context variable so every run signs
// automatically. Signing stays fail-open. Returns the handler.
export async function enableLangchainGovernance(
  opts: AsqavCallbackHandlerOptions,
): Promise<AsqavCallbackHandler & BaseCallbackHandlerLike> {
  const handler = await AsqavCallbackHandler.create(opts);
  const { registerConfigureHook, setContextVariable } = await loadContext();
  // Register the hook once so a second call cannot add a duplicate entry to
  // langchain-core's hook list and double-sign every action.
  if (!hookRegistered) {
    registerConfigureHook({ contextVar: ASQAV_LC_CONTEXT_VAR, inheritable: true });
    hookRegistered = true;
  }
  setContextVariable(ASQAV_LC_CONTEXT_VAR, handler);
  return handler;
}

File: python/src/asqav/extras/langchain.py
"""LangChain integration for asqav.

Install: pip install asqav[langchain]

Default-on (one call, no per-invoke config)::

    from asqav.extras.langchain import enable_langchain_governance
    enable_langchain_governance(agent_name="my-langchain-agent")
    chain.invoke(input)  # signs receipts, no config={"callbacks": [...]}

Manual attach (unchanged, still supported)::

    from asqav.extras.langchain import AsqavCallbackHandler
    handler = AsqavCallbackHandler(agent_name="my-langchain-agent")
    chain.invoke(input, config={"callbacks": [handler]})
"""

from __future__ import annotations

from contextvars import ContextVar
from typing import Any

try:
    from langchain_core.callbacks import BaseCallbackHandler
    from langchain_core.outputs import LLMResult
except ImportError as err:
    raise ImportError(
        "LangChain integration requires langchain-core. "
        "Install with: pip install asqav[langchain]"
    ) from err

from ._base import AsqavAdapter

__all__ = ["AsqavCallbackHandler", "enable_langchain_governance"]


class AsqavCallbackHandler(BaseCallbackHandler, AsqavAdapter):  # type: ignore[misc]
    """LangChain callback handler that auto-signs chain, tool, and LLM events.

    Every chain run, tool call, and LLM interaction is signed via asqav
    for governance and audit. Signing failures are logged but never raise -
    governance must not break the user's AI pipeline.

    Args:
        api_key: Optional API key override (uses asqav.init() default).
        agent_name: Name for a new agent (calls Agent.create).
        agent_id: ID of an existing agent (calls Agent.get).
    """

    def __init__(
        self,
        *,
        api_key: str | None = None,
        agent_name: str | None = None,
        agent_id: str | None = None,
        observe: bool = False,
        **kwargs: Any,
    ) -> None:
        AsqavAdapter.__init__(
            self,
            api_key=api_key,
            agent_name=agent_name,
            agent_id=agent_id,
            observe=observe,
        )
        BaseCallbackHandler.__init__(self)

    # -- Chain callbacks -------------------------------------------------------

    def on_chain_start(
        self,
        serialized: dict[str, Any] | None,
        inputs: dict[str, Any],
        **kwargs: Any,
    ) -> None:
        """Sign chain:start with chain name and input keys.

        serialized may be None on the LCEL/RunnableLambda path; fall back
        to a stable "chain" label so the receipt is always signed.
        """
        # Guard: LCEL/RunnableLambda passes raw str/list/number as inputs.
        input_keys = list(inputs.keys()) if isinstance(inputs, dict) else []
        if not isinstance(serialized, dict):
            self._sign_action(
                "chain:start",
                {"chain": "chain", "input_keys": input_keys},
            )
            return
        chain_id = serialized.get("id", ["unknown"])
        name = serialized.get("name") or chain_id[-1]
        self._sign_action(
            "chain:start",
            {"chain": str(name), "input_keys": input_keys},
        )

    def on_chain_end(self, outputs: dict[str, Any], **kwargs: Any) -> None:
        """Sign chain:end with output keys."""
        self._sign_action("chain:end", {"output_keys": list(outputs.keys())})

    def on_chain_error(self, error: BaseException, **kwargs: Any) -> None:
        """Sign chain:error with error type and message."""
        self._sign_action(
            "chain:error",
            {"error_type": type(error).__name__, "error": str(error)[:200]},
        )

    # -- Tool callbacks --------------------------------------------------------

    def on_tool_start(
        self,
        serialized: dict[str, Any],
        input_str: str,
        **kwargs: Any,
    ) -> None:
        """Sign tool:start with tool name and truncated input."""
        name = serialized.get("name") or serialized.get("id", ["unknown"])[-1]
        self._sign_action(
            "tool:start",
            {"tool": str(name), "input": str(input_str)[:200]},
        )

    def on_tool_end(self, output: Any, **kwargs: Any) -> None:
        """Sign tool:end with output type and length."""
        output_str = str(output)
        self._sign_action(
            "tool:end",
            {"output_type": type(output).__name__, "output_length": len(output_str)},
        )

    def on_tool_error(self, error: BaseException, **kwargs: Any) -> None:
        """Sign tool:error with error details."""
        self._sign_action(
            "tool:error",
            {"error_type": type(error).__name__, "error": str(error)[:200]},
        )

    # -- LLM callbacks ---------------------------------------------------------

    def on_llm_start(
        self,
        serialized: dict[str, Any],
        prompts: list[str],
        **kwargs: Any,
    ) -> None:
        """Sign llm:start with model name and prompt count."""
        model = serialized.get("name") or serialized.get("id", ["unknown"])[-1]
        self._sign_action(
            "llm:start",
            {"model": str(model), "prompt_count": len(prompts)},
        )

    def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:
        """Sign llm:end with generation count and token usage if available."""
        generation_count = sum(len(g) for g in response.generations)
        context: dict[str, Any] = {"generation_count": generation_count}

        if response.llm_output and isinstance(response.llm_output, dict):
            token_usage = response.llm_output.get("token_usage")
            if token_usage and isinstance(token_usage, dict):
                context["token_usage"] = {
                    k: v for k, v in token_usage.items() if isinstance(v, (int, float))
                }

        self._sign_action("llm:end", context)

    def on_llm_error(self, error: BaseException, **kwargs: Any) -> None:
        """Sign llm:error with error details."""
        self._sign_action(
            "llm:error",
            {"error_type": type(error).__name__, "error": str(error)[:200]},
        )


# -- Default-on entrypoint -----------------------------------------------------

# LangChain auto-attaches a handler to every run when its ContextVar is set and
# the paired configure hook is registered (langchain_core _configure loop).
_ASQAV_LC_HANDLER: ContextVar[AsqavCallbackHandler | None] = ContextVar(
    "asqav_langchain_handler", default=None
)
_hook_registered = False


def enable_langchain_governance(
    *,
    api_key: str | None = None,
    agent_name: str | None = None,
    agent_id: str | None = None,
    observe: bool = False,
) -> AsqavCallbackHandler:
    """Turn on default-on asqav governance for LangChain in one call.

    After this call every chain, tool, and LLM run in the current context
    signs an asqav receipt with no per-invoke ``config={"callbacks": [...]}``.
    It registers a langchain_core configure hook once and binds the handler to
    a ContextVar; LangChain's run configuration then attaches it to every run.
    Signing stays fail-open and the returned handler holds the signatures.
    """
    global _hook_registered
    from langchain_core.tracers.context import register_configure_hook

    handler = AsqavCallbackHandler(
        api_key=api_key,
        agent_name=agent_name,
        agent_id=agent_id,
        observe=observe,
    )
    # Register the configure hook at most once so a second call cannot
    # attach the ContextVar handler twice per run (double-sign).
    if not _hook_registered:
        register_configure_hook(_ASQAV_LC_HANDLER, True)
        _hook_registered = True
    _ASQAV_LC_HANDLER.set(handler)
    return handler
